core/vcpu.c

```diff
@@ -2067,551 +1957,175 @@ static int vcpu_simple_decode(struct vcpu_t *vcpu, struct decode *dc)
+    switch (segment) {
+    case SEG_CS:
+        return vcpu->state->_cs.base;
+    case SEG_DS:
+        return vcpu->state->_ds.base;
+    case SEG_ES:
+        return vcpu->state->_es.base;
+    case SEG_FS:
+        return vcpu->state->_fs.base;
+    case SEG_GS:
+        return vcpu->state->_gs.base;
+    case SEG_SS:
+        return vcpu->state->_ss.base;
+    default:
+        return vcpu->state->_ds.base;
+    }
+}
 
-            dc->opcode_dir = OPCODE_AND_IOMEM_TO_REG;
-            dc->size = instr[num] == 0x22 ? 1 : operand_size;
-            dc->reg_index = modrm.reg + (rex_r ? 8 : 0);
-            break;
-        }
-        // TODO: Handle 0x08, 0x09, 0x80 (/1), 0x81 (/1) and 0x83 (/1)
-        case 0x0a:    // OR reg/mem => reg, 8-bit
-        case 0x0b: {  // OR reg/mem => reg, 16/32/64-bit
-            if (modrm.mod == 3) {
-                hax_error("Unexpected reg-to-reg OR instruction\n");
-                goto case_unexpected_opcode;
-            }
+static void vcpu_advance_rip(void *obj, uint64_t len)
+{
+    struct vcpu_t *vcpu = obj;
+    advance_rip(vcpu);
+}
 
-            dc->opcode_dir = OPCODE_OR_IOMEM_TO_REG;
-            dc->size = instr[num] == 0x0a ? 1 : operand_size;
-            dc->reg_index = modrm.reg + (rex_r ? 8 : 0);
-            break;
-        }
-        // TODO: Handle 0x30, 0x31, 0x80 (/6), 0x81 (/6) and 0x83 (/6)
-        case 0x32:    // XOR reg/mem => reg, 8-bit
-        case 0x33: {  // XOR reg/mem => reg, 16/32/64-bit
-            if (modrm.mod == 3) {
-                hax_error("Unexpected reg-to-reg XOR instruction\n");
-                goto case_unexpected_opcode;
-            }
+static em_status_t vcpu_read_memory(void *obj, uint64_t ea, uint64_t *value,
+                                    uint32_t size, uint32_t flags)
+{
+    struct vcpu_t *vcpu = obj;
+    uint64_t pa;
 
-            dc->opcode_dir = OPCODE_XOR_IOMEM_TO_REG;
-            dc->size = instr[num] == 0x32 ? 1 : operand_size;
-            dc->reg_index = modrm.reg + (rex_r ? 8 : 0);
-            break;
-        }
-case_unexpected_opcode:
-        default: {
-            hax_panic_vcpu(vcpu, "Unexpected MMIO instruction (opcode=0x%x,"
-                           " exit_instr_length=%u, num=%d, gpa=0x%llx,"
-                           " instr[0..5]=0x%x 0x%x 0x%x 0x%x 0x%x 0x%x)\n",
-                           instr[num], vcpu->vmx.exit_instr_length, num,
-                           dc->gpa, instr[0], instr[1], instr[2], instr[3],
-                           instr[4], instr[5]);
-            dump_vmcs(vcpu);
-            return -1;
-        }
+    if (flags & EM_OPS_NO_TRANSLATION) {
+        pa = vmread(vcpu, VM_EXIT_INFO_GUEST_PHYSICAL_ADDRESS);
+    } else {
+        vcpu_translate(vcpu, ea, 0, &pa, NULL, false);
     }
 
-    // Calculate instruction length
-    len = (uint8)(num + 1);  // 1 for opcode
-    if (has_modrm) {
-        len++;
-        if (has_sib) {
-            len++;
+    if (is_mmio_address(vcpu, pa)) {
```

Comment:
nit: When we skip translation, we should skip this check as well, or assert it is true, because we assume the GPA is an MMIO address.